frame/vesting/src/lib.rs

```diff
@@ -300,72 +359,127 @@ impl<T: Config> VestedTransfer for Pallet<T> {
 			asset,
 			schedule,
+			vesting_schedule_id,
 		});
 
 		Ok(())
 	}
 }
 
 impl<T: Config> Pallet<T> {
-	fn do_claim(who: &AccountIdOf<T>, asset: AssetIdOf<T>) -> Result<BalanceOf<T>, DispatchError> {
-		let locked = Self::locked_balance(who, asset);
+	fn do_claim(
+		who: &AccountIdOf<T>,
+		asset: AssetIdOf<T>,
+		vesting_schedule_ids: VestingScheduleIdSet<T::VestingScheduleId, T::MaxVestingSchedules>, /* Option<T::VestingScheduleId>, */
+	) -> Result<BalanceOf<T>, DispatchError> {
+		let locked = Self::locked_balance(who, asset, vesting_schedule_ids)?;
+
 		if locked.is_zero() {
 			// cleanup the storage and unlock the fund
 			<VestingSchedules<T>>::remove(who, asset);
 			T::Currency::remove_lock(VESTING_LOCK_ID, asset, who)?;
 		} else {
 			T::Currency::set_lock(VESTING_LOCK_ID, asset, who, locked)?;
 		}
+
 		Ok(locked)
 	}
 
-	/// Returns locked balance based on current block number.
-	fn locked_balance(who: &AccountIdOf<T>, asset: AssetIdOf<T>) -> BalanceOf<T> {
-		<VestingSchedules<T>>::mutate_exists(who, asset, |maybe_schedules| {
-			let total = if let Some(schedules) = maybe_schedules.as_mut() {
-				let mut total: BalanceOf<T> = Zero::zero();
-				schedules.retain(|s| {
-					let amount = s.locked_amount(
-						frame_system::Pallet::<T>::current_block_number(),
-						T::Time::now(),
-					);
-					total = total.saturating_add(amount);
-					!amount.is_zero()
-				});
-				total
-			} else {
-				Zero::zero()
-			};
-			if total.is_zero() {
-				*maybe_schedules = None;
+	/// Returns total locked balance and balance per vesting schedule, based on current block number
+	fn locked_balance(
+		who: &AccountIdOf<T>,
+		asset: AssetIdOf<T>,
+		vesting_schedule_ids: VestingScheduleIdSet<T::VestingScheduleId, T::MaxVestingSchedules>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		<VestingSchedules<T>>::try_mutate_exists(who, asset, |maybe_schedules| {
+			match maybe_schedules {
+				Some(schedules) => {
+					let ids_to_claim: Vec<_> = match vesting_schedule_ids {
+						VestingScheduleIdSet::All => schedules.keys().copied().collect(),
+						VestingScheduleIdSet::Many(ids) => ids.into_inner(),
+						VestingScheduleIdSet::One(id) => vec![id],
+					};
+
+					let mut ids_to_remove = vec![];
+
+					let mut total_locked_amount: BalanceOf<T> = Zero::zero();
+
+					for id_to_claim in ids_to_claim {
+						let schedule = schedules
+							.get_mut(&id_to_claim)
+							.ok_or(Error::<T>::VestingScheduleNotFound)?;
+						let locked_amount = schedule.locked_amount(
+							frame_system::Pallet::<T>::current_block_number(),
+							T::Time::now(),
+						);
+
+						let total_amount = schedule.total_amount()?;
+						let available_amount = total_amount.safe_sub(&locked_amount)?;
+
+						// if the schedule id is specified, we update the claimed
+						// amount for the specified schedule
+						schedule.already_claimed = available_amount;
+
+						if locked_amount.is_zero() {
+							ids_to_remove.push(id_to_claim);
```

Comment:
Why not just remove the `id_to_claim` from schedules here?

Reply:
It's a bit tricky, because if we remove it here, then we lose track of the `already_claimed` balance for this schedule, and can not use it on the next loop. We can also accumulate all the balances from the removed schedules before removing and then subtract this from the `total_locked_amount`, but seems like the code would get harder to understand in the future.

Reply:
So the goal seems to be to get the `total_locked_amount`. If I understand correctly the calculation for it would still be correct without the removed schedules as sum of `total_locked_amount` for removed ones is 0.

Reply:
You are actually right. A previous implementation didn't allow me to do this, but the current one does. Well spotted! 👁️